File: backend/models/transaction.model.js
const mongoose = require('mongoose');

const transactionSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'User ID is required']
  },
  type: {
    type: String,
    enum: ['topUp', 'gameTransaction'],
    required: [true, 'Transaction type is required']
  },
  amount: {
    type: Number,
    required: [true, 'Transaction amount is required']
  },
  description: {
    type: String,
    required: [true, 'Transaction description is required']
  },
  gameId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Game',
    default: null
  },
  timestamp: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true
});

// Index for efficient querying
transactionSchema.index({ userId: 1, timestamp: -1 });

// Virtual property to get rupiah amount
transactionSchema.virtual('rupiahAmount').get(function() {
  return this.amount * 500; // 1 chip = 500 rupiah
});

// Method to create a top-up transaction
transactionSchema.statics.createTopUp = async function(userId, chipAmount) {
  return this.create({
    userId,
    type: 'topUp',
    amount: chipAmount,
    description: 'Top-up',
  });
};

// Method to create a game transaction
transactionSchema.statics.createGameTransaction = async function(
  userId,
  chipAmount,
  gameId,
  description = 'Game transaction'
) {
  return this.create({
    userId,
    type: 'gameTransaction',
    amount: chipAmount,
    description: description,
    gameId: gameId
  });
};

// Method to get user's transaction history
transactionSchema.statics.getUserTransactions = async function(userId, limit = 10, page = 1) {
  const skip = (page - 1) * limit;

  return this.find({ userId })
    .sort({ timestamp: -1 })
    .skip(skip)
    .limit(limit)
    .populate('gameId', 'name')
    .exec();
};

const Transaction = mongoose.model('Transaction', transactionSchema);

module.exports = Transaction;

File: frontend/lib/models/transaction_model.dart
enum TransactionType {
  topUp,
  gameTransaction,
}

class Transaction {
  final String id;
  final TransactionType type;
  final int amount;
  final DateTime timestamp;
  final String description;
  final String? gameId;

  Transaction({
    required this.id,
    required this.type,
    required this.amount,
    required this.timestamp,
    required this.description,
    this.gameId,
  });

  // Factory constructor to create a Transaction from JSON
  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'],
      type: TransactionType.values.byName(json['type']),
      amount: json['amount'],
      timestamp: DateTime.parse(json['timestamp']),
      description: json['description'],
      gameId: json['gameId'],
    );
  }

  // Convert Transaction to JSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'type': type.name,
      'amount': amount,
      'timestamp': timestamp.toIso8601String(),
      'description': description,
      'gameId': gameId,
    };
  }
}